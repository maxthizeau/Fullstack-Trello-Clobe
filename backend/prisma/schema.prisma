generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id           Int           @id @default(autoincrement())
  publicId     String        @unique
  name         String        @db.VarChar(255)
  registeredAt DateTime      @default(now())
  updatedAt    DateTime      @updatedAt
  email        String        @unique
  password     String
  role         Role          @default(USER)
  boards       Board[]
  teams        UsersOnTeam[]
}

model Team {
  id      Int           @id @default(autoincrement())
  name    String
  boards  Board[]
  members UsersOnTeam[]
}

model UsersOnTeam {
  id      Int     @id @default(autoincrement())
  user    User    @relation(fields: [userId], references: [id])
  userId  Int
  team    Team    @relation(fields: [teamId], references: [id])
  teamId  Int
  isAdmin Boolean
}

model Task {
  id          Int       @id @default(autoincrement())
  name        String
  description String?
  checked     Boolean   @default(false)
  order       Int       @default(1)
  taskGroupId Int
  taskGroup   TaskGroup @relation(fields: [taskGroupId], references: [id], onDelete: Cascade)
}

model TaskGroup {
  id          Int     @id @default(autoincrement())
  name        String
  description String?
  boardId     Int
  order       Int     @default(1)
  board       Board   @relation(fields: [boardId], references: [id], onDelete: Cascade)
  tasks       Task[]
}

model Board {
  id          Int         @id @default(autoincrement())
  name        String
  description String?
  ownerId     Int
  teamId      Int?
  owner       User        @relation(fields: [ownerId], references: [id])
  team        Team?       @relation(fields: [teamId], references: [id])
  taskGroups  TaskGroup[]
}

enum Role {
  USER
  ADMIN
}
