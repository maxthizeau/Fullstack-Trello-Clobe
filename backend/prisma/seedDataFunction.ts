import bcrypt from "bcryptjs"

export const usersCredentials = {
  admin: { email: "john.admin@seed.com", password: "Passwd01" },
  bob: { email: "bob.user@seed.com", password: "Passwd02" },
  max: { email: "mac.user@seed.com", password: "Passwd03" },
  michael: { email: "michael.user@seed.com", password: "Passwd04" },
}

export async function seedData(prisma) {
  // Adding Users :

  const johnAdmin = await prisma.user.upsert({
    where: { email: "john.admin@seed.com" },
    update: {},
    create: {
      email: "john.admin@seed.com",
      name: "JohnAdmin",
      publicId: "JohnAdmin#12345",
      password: await bcrypt.hash("Passwd01", 10),
      role: "ADMIN",
    },
  })

  const bobUser = await prisma.user.upsert({
    where: { email: "bob.user@seed.com" },
    update: {},
    create: {
      email: "bob.user@seed.com",
      name: "BobUser",
      publicId: "BobUser#56789",
      password: await bcrypt.hash("Passwd02", 10),
      role: "USER",
    },
  })

  const maxUser = await prisma.user.upsert({
    where: { email: "max.user@seed.com" },
    update: {},
    create: {
      email: "max.user@seed.com",
      name: "MaxUser",
      publicId: "MaxUser#13579",
      password: await bcrypt.hash("Passwd03", 10),
      role: "USER",
    },
  })

  const michaelUser = await prisma.user.upsert({
    where: { email: "michael.user@seed.com" },
    update: {},
    create: {
      email: "michael.user@seed.com",
      name: "MichaelUser",
      publicId: "MichaelUser#24680",
      password: await bcrypt.hash("Passwd04", 10),
      role: "USER",
    },
  })

  console.log("üëç users added to database")
  //   console.log({ johnAdmin, bobUser, maxUser, michaelUser })

  // Creating Teams - For each teams, create a board, a taskGroup and 2 tasks
  //   1. One team with Bob & Max
  const teamBobMax = await prisma.team.create({
    data: {
      name: "Bob & Max team",
      members: {
        create: [
          { userId: bobUser.id, isAdmin: true },
          { userId: maxUser.id, isAdmin: false },
        ],
      },
      boards: {
        create: {
          name: "Board of Bob&Max team",
          description: "To do lists of Bob&Max project",
          ownerId: bobUser.id,
          taskGroups: {
            create: [
              {
                name: "First Task Group of Bob&Max",
                description: "Will add description later",
                tasks: { create: [{ name: "First Task" }, { name: "Second Task" }] },
              },
              {
                name: "Second Task Group of Bob&Max",
                description: "We will also add description later",
                tasks: { create: [{ name: "Fourth Task" }, { name: "Fifth Task" }] },
              },
            ],
          },
        },
      },
    },
  })
  //   2. One team with Michael & Max
  const teamMichaelMax = await prisma.team.create({
    data: {
      name: "Michael & Max team",
      members: {
        create: [
          { userId: michaelUser.id, isAdmin: true },
          { userId: maxUser.id, isAdmin: false },
        ],
      },
      boards: {
        create: {
          name: "Board of Michael&Max team",
          description: "To do lists of Michael&Max project",
          ownerId: michaelUser.id,
          taskGroups: {
            create: [
              {
                name: "First Task Group of Michael&Max",
                description: "This will be added later",
                tasks: { create: [{ name: "First Task" }, { name: "Second Task" }] },
              },
              {
                name: "Second Task Group of Michael&Max",
                description: "To be added",
                tasks: { create: [{ name: "Fourth Task" }, { name: "Fifth Task" }] },
              },
            ],
          },
        },
      },
    },
  })
  //   3. One team with Bob & Michael
  const teamBobMichael = await prisma.team.create({
    data: {
      name: "Bob & Michael team",
      members: {
        create: [
          { userId: bobUser.id, isAdmin: true },
          { userId: michaelUser.id, isAdmin: false },
        ],
      },
      boards: {
        create: {
          name: "Board of Bob & Michael team",
          description: "To do lists of Bob & Michael project",
          ownerId: bobUser.id,
          taskGroups: {
            create: [
              {
                name: "First Task Group of Bob & Michael",
                description: "This will be added later",
                tasks: { create: [{ name: "First Task" }, { name: "Second Task" }] },
              },
              {
                name: "Second Task Group of Bob & Michael",
                description: "To be added",
                tasks: { create: [{ name: "Fourth Task" }, { name: "Fifth Task" }] },
              },
            ],
          },
        },
      },
    },
  })

  // 4. Create a single board with no team but with bob as the owner
  const bobBoard = await prisma.board.create({
    data: {
      name: "Only Bob can access this board",
      description: "Not owned by any team, owner is bob",
      ownerId: bobUser.id,
      taskGroups: {
        create: {
          name: "Bob Task Group",
          tasks: {
            create: {
              name: "One of many thing to do",
            },
          },
        },
      },
    },
  })

  console.log("üëç Teams, boards, tasksgroups, tasks added to database")
  //   console.log({ teamBobMax, teamMichaelMax, teamBobMichael, bobBoard })
}

export async function deleteAll(prisma, log = false) {
  const deletedUserOnTeamCount = await prisma.usersOnTeam.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedUserOnTeamCount.count} users on team relations`)
  const deletedTasksCount = await prisma.task.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedTasksCount.count} tasks`)
  const deletedTaskGroupsCount = await prisma.taskGroup.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedTaskGroupsCount.count} task groups`)
  const deletedBoardsCount = await prisma.board.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedBoardsCount.count} boards`)
  const deletedTeamsCount = await prisma.team.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedTeamsCount.count} teams`)
  const deletedUserCount = await prisma.user.deleteMany({})
  log && console.log(`üóëÔ∏è Deleted ${deletedUserCount.count} users`)
}
