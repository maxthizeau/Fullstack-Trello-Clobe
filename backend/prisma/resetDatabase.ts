import { PrismaClient } from "@prisma/client"
import { deleteAll } from "./seedDataFunction"

const prisma = new PrismaClient()

deleteAll()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
