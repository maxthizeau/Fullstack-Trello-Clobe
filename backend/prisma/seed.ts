import { seedData } from "./seedDataFunction"
import { PrismaClient } from "@prisma/client"

const prisma = new PrismaClient()

seedData(prisma)
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
