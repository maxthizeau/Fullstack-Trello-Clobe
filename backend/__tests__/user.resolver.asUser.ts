import { PrismaClient } from "@prisma/client"
import { gql } from "apollo-server-express"
import { toPromise } from "apollo-link"
import { constructTestServer, startTestServer } from "./__utils"
import { deleteAll, seedData, usersCredentials } from "../prisma/seedDataFunction"
const TEST_PORT = 6601
const prisma = new PrismaClient()

// Mutation used in BeforeAll to get token auth
const LOGIN_MUTATION = gql`
  mutation login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      token
      user {
        id
      }
    }
  }
`

describe("User queries as logged in user", () => {
  let stop, graphql
  let token = ""

  // Before all test, insert seed data, create a temporary server used to log in and get Auth token.
  beforeAll(async () => {
    // Seed Data :
    const resultSeedData = await seedData(prisma).catch((e) => {
      console.error(e)
    })

    console.log("âœ¨ Seed data successfully created")

    // Create the temporary server and request login mutation
    const serverForLogin = constructTestServer(prisma)
    let testServer = await startTestServer(serverForLogin, TEST_PORT)
    const resultTest = await toPromise(
      testServer.graphql({
        query: LOGIN_MUTATION,
        variables: { email: usersCredentials.bob.email, password: usersCredentials.bob.password },
      })
    )
    token = resultTest.data?.login?.token
    // Stop the tmp server
    testServer.stop()
  })

  afterAll(async () => {
    // Delete all data after tests
    const resultDeleteAll = await deleteAll(prisma)
    await prisma.$disconnect()
  })
  beforeEach(async () => {
    // Create a server and pass the auth token
    const server = constructTestServer(prisma)
    const testServer = await startTestServer(server, TEST_PORT, token)
    stop = testServer.stop
    graphql = testServer.graphql
  })
  afterEach(async () => {
    // Stop the server after a test
    stop()
  })

  it("Query allUsers should return the list of users when logged in as basic user", async () => {
    const res: any = await toPromise(
      graphql({
        query: gql`
          query allUsers {
            allUsers {
              id
              email
              name
            }
          }
        `,
      })
    )
    expect(res).toMatchSnapshot()
  })
})
