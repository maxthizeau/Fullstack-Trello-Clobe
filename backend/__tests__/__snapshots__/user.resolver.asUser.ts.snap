// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User queries as logged in user Query allUsers should return the list of users when logged in as basic user 1`] = `
Object {
  "data": Object {
    "allUsers": Array [
      Object {
        "email": "john.admin@seed.com",
        "id": 1,
        "name": "JohnAdmin",
      },
      Object {
        "email": "bob.user@seed.com",
        "id": 2,
        "name": "BobUser",
      },
      Object {
        "email": "max.user@seed.com",
        "id": 3,
        "name": "MaxUser",
      },
      Object {
        "email": "michael.user@seed.com",
        "id": 4,
        "name": "MichaelUser",
      },
      Object {
        "email": "this.is@my.email",
        "id": 5,
        "name": "NewUser",
      },
    ],
  },
}
`;
